Add SymbolHelper::FindObjectInCache and VerifyObjectFile

dpallotti opened the pull request:
`SymbolHelper::FindObjectInCache` looks for the module itself, so it checks both the build-id and the size. The corresponding `VerifyObjectFile` is added as well, which checks for both build id and file size as we want an exact match.

This will be used in the context of the "fallback symbols": when symbols are not available, we cache the module itself. This method will check whether the module is already in the cache (it doesn't need to have symbols) after the whole search for symbols has completed (including in the cache).

Bug: http://b/245680119

Test:
- New unit tests.
- As part of the larger symbol loading fallback prototype.

dpallotti commented on src/Symbols/SymbolHelper.cpp:
```diff
@@ -266,10 +264,41 @@ ErrorMessageOr<fs::path> SymbolHelper::FindSymbolsInCache(const fs::path& module
+      [&build_id,
+       expected_file_size](const std::filesystem::path& cache_file_path) -> ErrorMessageOr<void> {
+        OUTCOME_TRY(VerifyFileSize(cache_file_path, expected_file_size));
+        return VerifyObjectFile(cache_file_path, build_id);
```
Since we want an exact match, checking the build id is extra safety. If we didn't, this change would simplify a bit because `VerifyObjectFile` wouldn't be completely necessary. Let me know what you think.

antonrohr replied:
I think I would leave it in. Orbit always requires the build id currently and only Mizar uses the file size. Only relying on file size feels a bit risky to me (even though it is probably fine in reality). I would rather improve the cache mechanism completely to support files without build id

dpallotti replied:
Sounds good!

antonrohr commented on src/Symbols/include/Symbols/SymbolUtils.h:
```diff
@@ -26,11 +26,15 @@ namespace orbit_symbols {
-// Checks if the file at symbol_file_path can be read as symbol file (elf, coff, pdb) and compares
-// the size of the file with expected_file_size. Returns void if the sizes are the same ErrorMessage
+// Checks if the file at object_file_path can be read as an object file (elf, coff) and compares the
+// build id of the file with module_build_id. Returns void if build ids are the same, ErrorMessage
```
```suggestion
// build id of the file with build_id. Returns void if build ids are the same, ErrorMessage
```

dpallotti replied:
I actually fixed the parameter names across a couple of functions as they were inconsistent (including between declaration and definition).

antonrohr reviewed (commented):
This is a really well executed change. I do not have any remarks on a code level. Thank you for also incorporating all these small changes to existing code. 

I have one high level remark. The functionality with checking file sizes is currently only used in Mizar. And Mizar makes an implicit assumption that everything in the cache folder also contains symbols, which will not be true anymore soon. 

There are multiple options of dealing with this. The best is probably for Mizar to support fallback symbols, but this does not make sense in this PR. Maybe `FindSymbolsInCache(const std::filesystem::path& module_path, uint64_t expected_file_size)` could not just verify the size, but also that the file actually contains symbols.

@akopich for visibility

dpallotti commented:
> Maybe `FindSymbolsInCache(const std::filesystem::path& module_path, uint64_t expected_file_size)` could not just verify the size, but also that the file actually contains symbols.

This is a good idea! I had renamed the file size version of `VerifySymbolsFile` to `VerifyFileSize` because it had nothing to do with symbols, but the way to go is for that function to verify the presence of symbols, just like the other version of `VerifySymbolsFile`: it checks symbols, not only the build id.

I'll do that and come back.

vickyliu-go4it reviewed (approved):
Looks very good from my side.

dpallotti commented:
> > Maybe `FindSymbolsInCache(const std::filesystem::path& module_path, uint64_t expected_file_size)` could not just verify the size, but also that the file actually contains symbols.
> 
> This is a good idea! I had renamed the file size version of `VerifySymbolsFile` to `VerifyFileSize` because it had nothing to do with symbols, but the way to go is for that function to verify the presence of symbols, just like the other version of `VerifySymbolsFile`: it checks symbols, not only the build id.
> 
> I'll do that and come back.

Ok, so now the file-size version of `VerifySymbolFile` also tries `CreateSymbolsFile`, which checks for symbols. Also I moved the size verification of `FindObjectInCache` to `VerifyObjectFile`.

antonrohr commented on src/Symbols/include/Symbols/SymbolUtils.h:
```diff
@@ -22,14 +22,20 @@ namespace orbit_symbols {
 // otherwise.
 [[nodiscard]] ErrorMessageOr<void> VerifySymbolFile(const std::filesystem::path& symbol_file_path,
+                                                    const std::string& build_id);
+// CChecks if the file at symbol_file_path can be read as symbol file (elf, coff, pdb) and compares
```
```suggestion
// Checks if the file at symbol_file_path can be read as symbol file (elf, coff, pdb) and compares
```

dpallotti replied:
Done.

antonrohr reviewed (approved):
LGTM!
